pkg/eval/unix/umask.go

```diff
@@ -0,0 +1,94 @@
+// +build !windows,!plan9
+
+package unix
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"syscall"
+
+	"github.com/elves/elvish/pkg/eval"
+	"github.com/elves/elvish/pkg/eval/vars"
+)
+
+//elvdoc:var umask
+//
+// The file system creation mask. When used in a string context the output is
+// its Elvish octal representation; e.g. 0o027. This makes it possible to use
+// it in any context that expects a `$number`.
+//
+// When assigning a new value a string is implicitly treated as an octal
+// number. If that fails the usual rules for interpreting
+// [numbers](./language.html#number) are used. This means you can do both
+// `unix:umask = 027` and `unix:umask = 0o27`. You can also assign to it a
+// `float64` data type that has no fractional component. The legal values are
+// in the range [0 .. 0o777].
+//
+// You can do a temporary assignment to affect a single command; e.g.
+// `umask=077 touch a_file`. After the command completes the old umask will be
+// restored. **Warning**: Since the umask applies to the entire process, not
+// individual threads, changing it temporarily in this manner is dangerous if
+// you are doing anything in parallel. Such as via the
+// [`peach`](ref/builtin.html#peach) command.
+//
+// This variable only has meaning on UNIX like operating systems and does not
+// exist on other platforms. Do not use this in elvish scripts meant to work
+// on non-UNIX compatible systems.
+
+// UmaskVariable is a variable whose value always reflects the current file
+// creation permission mask. Setting it changes the current file creation
+// permission mask for the process (not an individual thread).
+type UmaskVariable struct {
+	ev *eval.Evaler
+}
+
+var _ vars.Var = UmaskVariable{}
+
+// Get returns the current file creation umask as a string.
+func (UmaskVariable) Get() interface{} {
+	// Note: The seemingly redundant syscall is because the syscall.Umask() API
```

Comment:
Interesting, I never knew this weird restriction of UNIX syscall API.

The workaround is fine per se, but now you have a race condition: if another thread calls `Umask` between the two `Umask` call here, this code will revert that change.

You can use a sync.RWMutex to guard the calls made in this file, although that still won't fix calls made in other places. Fortunately nowhere else in Elvish calls `Umask` (except for the daemon, but then the daemon does not run the Elvish interpreter).

Reply:
Yes, this is racy but that is inherent in the umask API. Even with a mutex you'll have a problem since that won't stop any code that uses the `umask()` syscall from racing with this method. Note that the fish shell (written in C++ and uses threads) does this without a mutex and it hasn't been a problem. A mutex here would provide a false sense of security against a race.